### PR TITLE
Replace Koalas with pandas API on Spark

### LINKED ISSUE #1864: Update koalas code to pyspark pandas API instead in Featuretools
The koalas library is now in maintenance mode only, and the core code has been integrated into pyspark, starting with version 3.2. We should update our koalas specific code to use the new pyspark pandas API: https://spark.apache.org/docs/latest/api/python/getting_started/quickstart_ps.html

### PR DESCRIPTION
**Pull request by jeff-hernandez**
Closes #1864

### CODE REVIEW

File: featuretools/utils/spark_utils.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ def replace_tuple_columns(pdf):
     return new_df
 
 
-def replace_nan_with_flag(pdf, flag=-1):
+def replace_nan_with_flag(pdf, flag=-1.):
```
Comment (jeff-hernandez) [inline featuretools/utils/spark_utils.py:15]: Spark series  doesn't support an array that contains floats and integers

```python
ps.from_pandas(pd.Series([[0.0, 0.0], [7.0, 3.0], [14.0, 6.0], [-1, -1], [-1, -1]]))
```

```
TypeError: element in array field 0: DoubleType can not accept object -1 in type <class 'int'>
```

File: featuretools/entityset/entityset.py [hunk 2]
```diff
@@ -1579,11 +1579,9 @@ def _get_or_create_index(index, make_index, df):
 
 
 def _create_index(df, index):
-    if isinstance(df, dd.DataFrame):
+    if isinstance(df, (dd.DataFrame, ps.DataFrame)):
         df[index] = 1
         df[index] = df[index].cumsum() - 1
```
Comment (jeff-hernandez) [inline featuretools/entityset/entityset.py:1584]: Spark doesn't support range type for column assignment

File: setup.cfg [hunk 3]
```diff
@@ -46,7 +46,7 @@ install_requires =
     dask[dataframe] >= 2021.10.0
     psutil >= 5.6.6
     click >= 7.0.0
-    woodwork >= 0.8.1
+    woodwork @ git+https://github.com/alteryx/woodwork.git@migrate-to-pyspark-api#egg=woodwork
```
Comment (jeff-hernandez) [inline setup.cfg:49]: We'll need to change this to the corresponding woodwork version before merging

**Comment by jeff-hernandez** [comment]
we'll need to switch the required unit tests from koalas to spark

File: docs/source/release_notes.rst [hunk 4]
```diff
@@ -228,7 +229,7 @@ v1.0.0 Oct 12, 2021
         * Update API Reference to match Featuretools 1.0 API (:pr:`1600`)
         * Update Index page to be Jupyter notebook and use Woodwork integration (:pr:`1602`)
         * Update Feature Descriptions page to be Jupyter notebook and use Woodwork integration (:pr:`1603`)
-        * Update Using Koalas EntitySets page to be Jupyter notebook and use Woodwork integration (:pr:`1604`)
```
Comment (rwedge) [inline docs/source/release_notes.rst:231]: I don't think we need to retroactively replace Koalas with Spark in the changelog

Comment (jeff-hernandez) [reply docs/source/release_notes.rst:231]: yep agreed, I missed that change possibly after merging conflicts locally

File: setup.cfg [hunk 5]
```diff
@@ -87,11 +87,10 @@ test =
     composeml >= 0.8.0
     urllib3 >= 1.26.5
     pyarrow >= 3.0.0
-koalas = 
-    pyspark >=3.0.0, <3.2.0
-    koalas >=1.8.1, <=1.8.2; python_version < '3.10'
+spark = 
+    pyspark >=3.2.0, <=3.2.1
     pandas >=1.3.0, <1.4.0
```
Comment (rwedge) [inline setup.cfg:92]: are we able to remove the pandas<1.4 constraint now?

File: spark-requirements.txt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+pyspark>=3.2.0,<3.2.1
```
Comment (rwedge) [inline spark-requirements.txt:1]: is this file still necessary?

File: setup.cfg (hunk 5 above)
Comment (jeff-hernandez) [reply setup.cfg:92]: not yet  there are still breaking changes

File: spark-requirements.txt (hunk 6 above)
Comment (jeff-hernandez) [reply spark-requirements.txt]: nope removed

File: setup.cfg (hunk 5 above)
Comment (rwedge) [reply setup.cfg:89]: are we tracking that issue somewhere?

Comment (jeff-hernandez) [reply setup.cfg:89]: it seems like it, opened #1951

File: featuretools/tests/entityset_tests/test_last_time_index.py [hunk 7]
```diff
@@ -180,8 +180,8 @@ def test_parent_no_time_index_missing(self, es, extra_session_df,
 
     def test_multiple_children(self, es, wishlist_df,
                                true_sessions_lti):
-        if ks and isinstance(es.dataframes[0], ks.DataFrame):
-            pytest.xfail('Cannot make index on a Koalas DataFrame')
+        if ps and isinstance(es.dataframes[0], ps.DataFrame):
```
Comment (thehomebrewnerd) [inline featuretools/tests/entityset_tests/test_last_time_index.py:183]: Can we update this to use the `if es.dataframe_type == Library.SPARK.value` approach that we are using elsewhere in this file? There are few locations in this file that could be updated.

File: .github/workflows/unit_tests_with_latest_deps.yml [hunk 8]
```diff
@@ -14,10 +14,10 @@ jobs:
       fail-fast: false
       matrix:
         python_version: ["3.7", "3.8", "3.9", "3.10"]
-        libraries: ["core", "koalas"]
+        libraries: ["core", "spark"]
         exclude:
           - python_version: "3.10"
-            libraries: "koalas"
+            libraries: "spark"
```
[the remark this answers is not in the document]
Comment (jeff-hernandez) [reply .github/workflows/unit_tests_with_latest_deps.yml]: updated

**Review by thehomebrewnerd — APPROVED** [review]
LGTM!

**Comment by jeff-hernandez** [comment]
I think we'll need to merge and release the changes in Woodwork first before doing the same in featuretools.

**Comment by thehomebrewnerd** [comment]
@jeff-hernandez Just a quick heads-up. The `compatibility` attribute for the primitives added by #1948 will need to be updated as well when you fix the release notes merge conflict.

File: setup.cfg [hunk 9]
```diff
@@ -84,11 +84,10 @@ test =
     composeml >= 0.8.0
     urllib3 >= 1.26.5
     pyarrow >= 3.0.0
-koalas = 
-    pyspark >=3.0.0, <3.2.0
-    koalas >=1.8.1, <=1.8.2; python_version < '3.10'
+spark = 
+    pyspark >=3.2.0, <=3.2.1
```
Comment (thehomebrewnerd) [inline setup.cfg:88]: Do we need an upper bound here?

Comment (jeff-hernandez) [reply setup.cfg]: not necessary, removed

**Review by thehomebrewnerd — APPROVED** [review]
LGTM, assuming tests pass.

**Pull request merged**